File: dns_lookup.php
<?php
// dns_lookup.php — PHP версия

class DNSLookupIPv6 {
    private $timeout;
    private $threads;
    private $results = [];

    public function __construct($timeout = 5, $threads = 20) {
        $this->timeout = $timeout;
        $this->threads = $threads;
    }

    private function isIPv6($addr) {
        return filter_var($addr, FILTER_VALIDATE_IP, FILTER_FLAG_IPV6) !== false;
    }

    private function lookupAAAA($domain) {
        $records = @dns_get_record($domain, DNS_AAAA);
        if ($records === false) return [];
        return array_column($records, 'ipv6');
    }

    private function lookupPTR($ip) {
        $ptr = @gethostbyaddr($ip);
        if ($ptr === false || $ptr === $ip) return [];
        return [$ptr];
    }

    private function lookupDomain($domain) {
        $domain = trim($domain);
        if (empty($domain)) return null;

        if ($this->isIPv6($domain)) {
            $ptr = $this->lookupPTR($domain);
            return [
                'domain' => $domain,
                'timestamp' => date('c'),
                'AAAA' => [],
                'PTR' => $ptr ?: ['не найден']
            ];
        }

        $aaaa = $this->lookupAAAA($domain);
        $ptr = [];
        if (!empty($aaaa)) {
            foreach ($aaaa as $ip) {
                $names = $this->lookupPTR($ip);
                $ptr = array_merge($ptr, $names);
            }
        }
        if (empty($ptr)) {
            $ptr = ['не найден'];
        }

        return [
            'domain' => $domain,
            'timestamp' => date('c'),
            'AAAA' => $aaaa,
            'PTR' => $ptr
        ];
    }

    public function lookupBatch($domains) {
        echo "🔍 Выполняем lookup для " . count($domains) . " доменов...\n";
        $start = microtime(true);

        foreach ($domains as $domain) {
            $result = $this->lookupDomain($domain);
            if ($result) {
                $this->results[] = $result;
                echo "✅ $domain — выполнено\n";
            }
        }

        $elapsed = microtime(true) - $start;
        echo "⏱️ Время выполнения: " . number_format($elapsed, 2) . " сек.\n";
        return $this->results;
    }

    public function saveJSON($filename) {
        file_put_contents($filename, json_encode($this->results, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE));
        echo "💾 Сохранено JSON: $filename\n";
    }

    public function saveCSV($filename) {
        if (empty($this->results)) return;
        $fp = fopen($filename, 'w');
        fputcsv($fp, ['Domain', 'Timestamp', 'AAAA', 'PTR']);
        foreach ($this->results as $r) {
            fputcsv($fp, [
                $r['domain'],
                $r['timestamp'],
                implode('; ', $r['AAAA']),
                implode('; ', $r['PTR'])
            ]);
        }
        fclose($fp);
        echo "💾 Сохранено CSV: $filename\n";
    }

    public function printTable() {
        if (empty($this->results)) {
            echo "Нет результатов.\n";
            return;
        }

        echo "\n" . str_repeat('=', 90) . "\n";
        printf("%-25s %-35s %-25s\n", "Домен/IP", "AAAA", "PTR");
        echo str_repeat('-', 90) . "\n";
        foreach ($this->results as $r) {
            $aaaa = implode('; ', $r['AAAA']);
            $aaaa = empty($aaaa) ? '—' : $aaaa;
            if (strlen($aaaa) > 35) $aaaa = substr($aaaa, 0, 35) . '...';
            $ptr = implode('; ', $r['PTR']);
            if (strlen($ptr) > 25) $ptr = substr($ptr, 0, 25) . '...';
            printf("%-25s %-35s %-25s\n", $r['domain'], $aaaa, $ptr);
        }
        echo str_repeat('=', 90) . "\n";
    }
}

function main($argv) {
    if ($argc < 2) {
        echo "Usage: php dns_lookup.php <domains.txt>\n";
        echo "   или: php dns_lookup.php domain1.com domain2.com\n";
        exit(1);
    }

    $domains = [];

    if ($argc == 2 && file_exists($argv[1])) {
        $domains = array_filter(array_map('trim', file($argv[1])), function($d) {
            return !empty($d);
        });
    } else {
        $domains = array_slice($argv, 1);
    }

    if (empty($domains)) {
        echo "❌ Нет доменов для проверки.\n";
        exit(1);
    }

    echo "🌐 DNS Lookup IPv6 Pro (PHP)\n";
    echo "📂 Загружено " . count($domains) . " доменов.\n";

    $lookup = new DNSLookupIPv6(5, 20);
    $lookup->lookupBatch($domains);

    if (!empty($lookup->results)) {
        $lookup->printTable();
        $lookup->saveJSON('dns_results.json');
        $lookup->saveCSV('dns_results.csv');
    } else {
        echo "❌ Не удалось выполнить lookup.\n";
    }
}

$argc = $_SERVER['argc'] ?? 0;
$argv = $_SERVER['argv'] ?? [];
main($argv);
?>

File: dns_lookup.java
// dns_lookup.java — Java версия

import java.io.*;
import java.nio.file.*;
import java.net.*;
import java.util.*;
import java.util.concurrent.*;
import java.time.*;
import org.xbill.DNS.*;

public class dns_lookup {
    private static final int TIMEOUT = 5000;
    private static final int THREADS = 20;
    private static List<Map<String, Object>> results = new ArrayList<>();

    public static void main(String[] args) throws Exception {
        if (args.length < 1) {
            System.out.println("Usage: java -cp .:dnsjava.jar dns_lookup <domains.txt>");
            System.out.println("   или: java -cp .:dnsjava.jar dns_lookup domain1.com domain2.com");
            System.exit(1);
        }

        List<String> domains = new ArrayList<>();

        if (args.length == 1 && new File(args[0]).exists()) {
            domains = Files.readAllLines(Paths.get(args[0]));
            domains.removeIf(String::isBlank);
        } else {
            domains = Arrays.asList(args);
        }

        if (domains.isEmpty()) {
            System.out.println("❌ Нет доменов для проверки.");
            System.exit(1);
        }

        System.out.println("🌐 DNS Lookup IPv6 Pro (Java)");
        System.out.println("📂 Загружено " + domains.size() + " доменов.");

        ExecutorService executor = Executors.newFixedThreadPool(THREADS);
        List<Future<Map<String, Object>>> futures = new ArrayList<>();

        long start = System.currentTimeMillis();

        for (String domain : domains) {
            futures.add(executor.submit(() -> lookupDomain(domain.trim())));
        }

        for (Future<Map<String, Object>> future : futures) {
            try {
                Map<String, Object> result = future.get();
                if (result != null && !result.isEmpty()) {
                    results.add(result);
                    System.out.println("✅ " + result.get("domain") + " — выполнено");
                }
            } catch (Exception e) {
                System.out.println("❌ Ошибка: " + e.getMessage());
            }
        }

        executor.shutdown();
        executor.awaitTermination(30, TimeUnit.SECONDS);

        double elapsed = (System.currentTimeMillis() - start) / 1000.0;
        System.out.printf("⏱️ Время выполнения: %.2f сек.\n", elapsed);

        if (!results.isEmpty()) {
            printTable();
            saveJSON("dns_results.json");
            saveCSV("dns_results.csv");
        } else {
            System.out.println("❌ Не удалось выполнить lookup.");
        }
    }

    @SuppressWarnings("unchecked")
    private static Map<String, Object> lookupDomain(String domain) {
        Map<String, Object> result = new LinkedHashMap<>();
        result.put("domain", domain);
        result.put("timestamp", Instant.now().toString());

        try {
            // Проверяем, является ли строка IPv6-адресом
            try {
                InetAddress addr = InetAddress.getByName(domain);
                if (addr instanceof Inet6Address) {
                    // Reverse lookup для IPv6
                    Name rev = ReverseMap.fromAddress(addr.getHostAddress());
                    Record[] ptrRecords = new Lookup(rev, Type.PTR).run();
                    List<String> ptr = new ArrayList<>();
                    if (ptrRecords != null) {
                        for (Record rec : ptrRecords) {
                            ptr.add(rec.rdataToString());
                        }
                    }
                    result.put("AAAA", new ArrayList<>());
                    result.put("PTR", ptr.isEmpty() ? Arrays.asList("не найден") : ptr);
                    return result;
                }
            } catch (UnknownHostException e) {
                // Не IP-адрес, продолжаем как домен
            }

            // AAAA записи
            Lookup lookup = new Lookup(domain, Type.AAAA);
            Record[] records = lookup.run();
            List<String> aaaa = new ArrayList<>();
            if (records != null) {
                for (Record rec : records) {
                    if (rec instanceof AAAARecord) {
                        aaaa.add(((AAAARecord) rec).getAddress().getHostAddress());
                    }
                }
            }
            result.put("AAAA", aaaa);

            // PTR для каждого AAAA адреса
            List<String> ptr = new ArrayList<>();
            if (!aaaa.isEmpty()) {
                for (String ip : aaaa) {
                    Name rev = ReverseMap.fromAddress(ip);
                    Record[] ptrRecords = new Lookup(rev, Type.PTR).run();
                    if (ptrRecords != null) {
                        for (Record rec : ptrRecords) {
                            ptr.add(rec.rdataToString());
                        }
                    }
                }
            }
            result.put("PTR", ptr.isEmpty() ? Arrays.asList("не найден") : ptr);

        } catch (Exception e) {
            result.put("AAAA", new ArrayList<>());
            result.put("PTR", Arrays.asList("не найден"));
        }

        return result;
    }

    @SuppressWarnings("unchecked")
    private static void printTable() {
        System.out.println("\n" + "=".repeat(90));
        System.out.printf("%-25s %-35s %-25s\n", "Домен/IP", "AAAA", "PTR");
        System.out.println("-".repeat(90));
        for (Map<String, Object> r : results) {
            String aaaa = String.join("; ", (List<String>) r.get("AAAA"));
            if (aaaa.isEmpty()) aaaa = "—";
            if (aaaa.length() > 35) aaaa = aaaa.substring(0, 35) + "...";
            String ptr = String.join("; ", (List<String>) r.get("PTR"));
            if (ptr.length() > 25) ptr = ptr.substring(0, 25) + "...";
            System.out.printf("%-25s %-35s %-25s\n", r.get("domain"), aaaa, ptr);
        }
        System.out.println("=".repeat(90));
    }

    @SuppressWarnings("unchecked")
    private static void saveJSON(String filename) throws IOException {
        String json = new com.google.gson.GsonBuilder().setPrettyPrinting().create().toJson(results);
        Files.write(Paths.get(filename), json.getBytes());
        System.out.println("💾 Сохранено JSON: " + filename);
    }

    @SuppressWarnings("unchecked")
    private static void saveCSV(String filename) throws IOException {
        if (results.isEmpty()) return;
        StringBuilder sb = new StringBuilder();
        sb.append("Domain,Timestamp,AAAA,PTR\n");
        for (Map<String, Object> r : results) {
            sb.append(r.get("domain")).append(",");
            sb.append(r.get("timestamp")).append(",");
            sb.append(String.join("; ", (List<String>) r.get("AAAA"))).append(",");
            sb.append(String.join("; ", (List<String>) r.get("PTR"))).append("\n");
        }
        Files.write(Paths.get(filename), sb.toString().getBytes());
        System.out.println("💾 Сохранено CSV: " + filename);
    }
}
